### DISCUSSION TITLE
C++ body_set_force_integration_callback, callable function parameters

### DISCUSSION THREAD

**Discussion by Mangonels** [godot-jolt/godot-jolt#965] [Questions] [answered]
Hello, I'm learning PhysicsServer3D with godot-cpp and godot-jolt.

**Asuming:**
- `body_set_force_integration_callback` can be used to set the position of a visual instance based on the body's current position from state (so the visual updates it's position by following the body kind of like child nodes would do in a scene tree).
- The userdata parameter in `body_set_force_integration_callback` is for "extra Variant data" to be sent to the Callable's second parameter instead of using `Callable::bind()` for that.

When constructing the **callback method** for `callable_mp`, as I understand it, the first argument should be the state (which is my main issue) and the second is a Variant with parameter data, here's how my callback method currently looks:

```c++
    void WorldModifier::on_pcc_body_update(const godot::Variant& state,
                                           const godot::RID& visual_instance_rid)
    {
        const godot::PhysicsDirectBodyState3D s = to<godot::PhysicsDirectBodyState3D>(state);
        gdeng::rendering_server()->instance_set_transform(visual_instance_rid, s.get_transform());
    }
```

I'm currently using `godot::Variant` as first parameter type (otherwise I get complaint that it's not Variant), however I also tried `godot::PhysicsDirectBodyState3D`, with errors indicating it needs to be a Variant type. Now it's a Variant, but my conversion is really lame and I can't get_transform.

**The request**: 
Ideally an example on how to construct my `callable_mp` and it's callback method for `body_set_force_integration_callback` would be amazing, but just letting me know what type would be accepted as first `state` parameter or how to convert from variant to an actually usable `PhysicsDirectBodyState3D` (if this is even posible which I doubt) would also be great!

**Comment by mihe** [comment, marked as answer]
> `body_set_force_integration_callback` can be used to set the position of a visual instance based on the body's current position from state

Keep in mind that using this callback for this purpose was only a somewhat necessary hack in Godot 4.2 and prior versions, but as of Godot 4.3 you can/should now use [`body_set_state_sync_callback`](https://docs.godotengine.org/en/stable/classes/class_physicsserver3d.html#class-physicsserver3d-method-body-set-state-sync-callback) instead.

> I'm currently using `godot::Variant` as first parameter type (otherwise I get complaint that it's not Variant), however I also tried `godot::PhysicsDirectBodyState3D`

It seems to work just fine for me if I use this signature:

```cpp
void MyClass::_my_method(const PhysicsDirectBodyState3D* p_state, const RID& p_visual_instance_rid);
```

... and connect it as such:

```cpp
PhysicsServer3D::get_singleton()->body_set_force_integration_callback(some_body->get_rid(), callable_mp(this, &MyClass::_my_method), some_visual_instance_rid);
```

Maybe the problem is that you were using a reference rather than a pointer?
